File: src/project_version.py
"""Load the shared project release version for the Python bridge.

The root VERSION file is the canonical release identifier used by the bridge,
gateway, dashboard, and future container images. It is read once at import time
so request handlers do not touch the filesystem.
"""

from __future__ import annotations

from pathlib import Path

VERSION_FILE = Path(__file__).resolve().parent.parent / "VERSION"


def _read_project_version(path: Path = VERSION_FILE) -> str:
    """Read and validate one non-empty project version from ``path``."""
    try:
        version = path.read_text(encoding="utf-8").strip()
    except FileNotFoundError as exc:
        raise RuntimeError(f"Project version file is missing: {path}") from exc

    if not version:
        raise RuntimeError(f"Project version file is empty: {path}")
    return version


APP_VERSION = _read_project_version()

File: fedify-gateway/src/project-version.ts
/** Load the shared project release version for the Fedify gateway. */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const VERSION_FILE = resolve(dirname(fileURLToPath(import.meta.url)), "../../VERSION");

/** Read one non-empty version value from the canonical root VERSION file. */
export function readProjectVersion(path = VERSION_FILE): string {
  let version: string;
  try {
    version = readFileSync(path, "utf8").trim();
  } catch (error) {
    throw new Error(`Project version file is missing: ${path}`, { cause: error });
  }

  if (version.length === 0) {
    throw new Error(`Project version file is empty: ${path}`);
  }
  return version;
}

export const APP_VERSION = readProjectVersion();
